### DISCUSSION TITLE
(WIP) RFC-022: support swc plugin

### DISCUSSION THREAD

**Discussion by jerrykingxyz** [web-infra-dev/rspack#733] [RFC]
- Feature Name: swc_plugin
- Start Date: 2022/09/16

# Summary
[summary]: #summary

how to use swc plugin in rspack

# Motivation
[motivation]: #motivation

the swc cannot satisfy custom transform feature, we need load some custom swc plugin to extension swc feature.
the custom plugin can from rspack or third-party like swc plugin which shared with modernjs

# Prior art
[prior-art]: #prior-art

### how to use swc plugin
there two way to use swc plugin
1. pack swc plugin to wasm file, use it in .swcrc `jsc.experimental.plugins`. [more info](https://github.com/swc-project/swc/discussions/3540)
this is official way to create swc plugin, it can dymantic install and use plugin. because of using wasm, the running speed is much better than js.
2. use swc plugin in rspack
this is the fastest way to use swc plugin, but it will make rspack larger

because of performance, we decided to use swc plugin in rspack and expose configurable options to control it.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### rspack swc plugin

the rspack swc plugin is officially developed and maintained by us, and its switches are in the rspack configuration file

``` javascript
// rspack config
{
  // define plugin
  "define": { "NULL": "null" }
}
```

### third-party swc plugins

the third-party plugin need change rspack code to support

```javascript
// rspack config
{
  "builtins": {
    "remove_console": true
  }
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

### rspack swc plugin

1. the swc plugin should be in the `visitors` folder
2. all uses of `visitor` are in `mod.rs`, and it expose those functions
* `before_pass`
* `after_pass`
* `scan_dependency`

``` rust
mod visitors;
// ...
ast.visit_with(visitors::xxx)
```

### third-party swc plugins

1. the third-party plugin should add as dependency in rspack
``` toml
[dependencies]
swc_plugin_remove_console = { git = "https://github.com/xx/xxx", package = "remove_console" }
# or in crates.io
swc_plugin_remove_console = "1.0.0"
```
4. use it in rspack and register `builtins.xxx` as config param

``` rust
if compilation.options.builtins.xxx {
  ast.visit_with(swc_plugin_remove_console::xxx);
};
```

### TODO

1. Optimize the use of `visitors` in current code, make it all in `mod.rs`
2. use modernjs swc plugin in rspack

# Drawbacks
[drawbacks]: #drawbacks

1. the third-party swc plugin if use git, we can not publish rspack to crates.io.(currently we have no plans to do this)
2. If the third party lib is private, we have no way to use it
5. packing all swc plugins into rspack will makes it bigger

**Comment by hardfist** [comment]
define also need to be put into builtins because webpack doesn't support this out of box

**Comment by hyf0** [comment]
How to deal with the conflict problem that the version of crate `swc` in `rspack` and `third_plugin` is inconsistent?
